File: flutter_inappwebview_forge_android/android/src/main/kotlin/com/emirkanacar/flutter_inappwebview_forge_android/find_interaction/FindInteractionChannelDelegate.kt
package com.emirkanacar.flutter_inappwebview_forge_android.find_interaction

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.HashMap
import com.emirkanacar.flutter_inappwebview_forge_android.types.ChannelDelegateImpl
import com.emirkanacar.flutter_inappwebview_forge_android.types.FindSession
import com.emirkanacar.flutter_inappwebview_forge_android.webview.in_app_webview.InAppWebView

open class FindInteractionChannelDelegate(
    private var findInteractionController: FindInteractionController?,
    channel: MethodChannel
) : ChannelDelegateImpl(channel) {

    private fun canDispatchCallbacks(): Boolean {
        val webView = findInteractionController?.webView
        return getChannel() != null && webView != null &&
            (webView !is InAppWebView || webView.acceptsCallbacks())
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (!canDispatchCallbacks()) {
            result.success(null)
            return
        }
        when (call.method) {
            "findAll" -> {
                findInteractionController?.findAll(call.argument<String>("find"))
                result.success(true)
            }
            "findNext" -> {
                findInteractionController?.findNext(call.argument<Boolean>("forward") ?: false)
                result.success(true)
            }
            "clearMatches" -> {
                findInteractionController?.clearMatches()
                result.success(true)
            }
            "setSearchText" -> {
                val controller = findInteractionController
                if (controller != null) {
                    controller.searchText = call.argument<String>("searchText")
                    result.success(true)
                } else {
                    result.success(false)
                }
            }
            "getSearchText" -> {
                result.success(findInteractionController?.searchText ?: false)
            }
            "getActiveFindSession" -> {
                result.success(findInteractionController?.activeFindSession?.toMap())
            }
            else -> result.notImplemented()
        }
    }

    fun onFindResultReceived(activeMatchOrdinal: Int, numberOfMatches: Int, isDoneCounting: Boolean) {
        if (!canDispatchCallbacks()) return
        val channel = getChannel() ?: return
        val controller = findInteractionController
        if (isDoneCounting && controller?.webView != null) {
            controller.activeFindSession = FindSession(numberOfMatches, activeMatchOrdinal)
        }

        val obj = HashMap<String, Any?>()
        obj["activeMatchOrdinal"] = activeMatchOrdinal
        obj["numberOfMatches"] = numberOfMatches
        obj["isDoneCounting"] = isDoneCounting
        channel.invokeMethod("onFindResultReceived", obj)
    }

    override fun dispose() {
        super.dispose()
        findInteractionController = null
    }
}

File: flutter_inappwebview_forge_ios/ios/flutter_inappwebview_forge_ios/Sources/flutter_inappwebview_forge_ios/FindInteraction/FindInteractionChannelDelegate.swift
//
//  FindInteractionChannelDelegate.swift
//  flutter_inappwebview
//
//  Created by Lorenzo Pichilli on 07/10/22.
//

import Foundation
import Flutter

public class FindInteractionChannelDelegate: ChannelDelegate {
    private weak var findInteractionController: FindInteractionController?
    
    public init(findInteractionController: FindInteractionController, channel: FlutterMethodChannel) {
        super.init(channel: channel)
        self.findInteractionController = findInteractionController
    }
    
    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard canDispatchCallbacks() else {
            result(nil)
            return
        }
        let arguments = call.arguments as? NSDictionary
        
        switch call.method {
            case "findAll":
                if let findInteractionController = findInteractionController {
                    let find = arguments!["find"] as! String
                    findInteractionController.findAll(find: find, completionHandler: {(value, error) in
                        if error != nil {
                            result(FlutterError(code: "FindInteractionChannelDelegate", message: error?.localizedDescription, details: nil))
                            return
                        }
                        result(true)
                    })
                } else {
                    result(false)
                }
                break
            case "findString":
                if let findInteractionController = findInteractionController {
                    let find = arguments!["find"] as! String
                    let caseSensitive = arguments!["caseSensitive"] as? Bool ?? false
                    let backwards = arguments!["backwards"] as? Bool ?? false
                    let wraps = arguments!["wraps"] as? Bool ?? true
                    findInteractionController.findString(
                        find: find,
                        caseSensitive: caseSensitive,
                        backwards: backwards,
                        wraps: wraps,
                        completionHandler: { found, error in
                            if error != nil {
                                result(FlutterError(code: "FindInteractionChannelDelegate", message: error?.localizedDescription, details: nil))
                                return
                            }
                            result(found)
                        }
                    )
                } else {
                    result(false)
                }
                break
            case "findNext":
                if let findInteractionController = findInteractionController {
                    let forward = arguments!["forward"] as! Bool
                    findInteractionController.findNext(forward: forward, completionHandler: {(value, error) in
                        if error != nil {
                            result(FlutterError(code: "FindInteractionChannelDelegate", message: error?.localizedDescription, details: nil))
                            return
                        }
                        result(true)
                    })
                } else {
                    result(false)
                }
                break
            case "clearMatches":
                if let findInteractionController = findInteractionController {
                    findInteractionController.clearMatches(completionHandler: {(value, error) in
                        if error != nil {
                            result(FlutterError(code: "FindInteractionChannelDelegate", message: error?.localizedDescription, details: nil))
                            return
                        }
                        result(true)
                    })
                } else {
                    result(false)
                }
                break
            case "setSearchText":
                if let findInteractionController = findInteractionController {
                    let searchText = arguments!["searchText"] as? String
                    findInteractionController.searchText = searchText
                    result(true)
                } else {
                    result(false)
                }
                break
            case "getSearchText":
                result(findInteractionController?.searchText)
                break
            case "isFindNavigatorVisible":
                if #available(iOS 16.0, *) {
                    if let interaction = findInteractionController?.webView?.findInteraction {
                        result(interaction.isFindNavigatorVisible)
                    } else {
                        result(false)
                    }
                } else {
                    result(false)
                }
                break
            case "updateResultCount":
                if #available(iOS 16.0, *) {
                    if let interaction = findInteractionController?.webView?.findInteraction {
                        interaction.updateResultCount()
                        result(true)
                    } else {
                        result(false)
                    }
                } else {
                    result(false)
                }
                break
            case "presentFindNavigator":
                if #available(iOS 16.0, *) {
                    if let interaction = findInteractionController?.webView?.findInteraction {
                        interaction.presentFindNavigator(showingReplace: false)
                        result(true)
                    } else {
                        result(false)
                    }
                } else {
                    result(false)
                }
                break
            case "dismissFindNavigator":
                if #available(iOS 16.0, *) {
                    if let interaction = findInteractionController?.webView?.findInteraction {
                        interaction.dismissFindNavigator()
                        result(true)
                    } else {
                        result(false)
                    }
                } else {
                    result(false)
                }
                break
            case "getActiveFindSession":
                if let findInteractionController = findInteractionController {
                    result(findInteractionController.activeFindSession?.toMap())
                } else {
                    result(nil)
                }
                break
            default:
                result(FlutterMethodNotImplemented)
                break
        }
    }
    
    public func onFindResultReceived(activeMatchOrdinal: Int, numberOfMatches: Int, isDoneCounting: Bool) {
        guard canDispatchCallbacks() else { return }
        if isDoneCounting, let findInteractionController = findInteractionController {
            findInteractionController.activeFindSession = FindSession(resultCount: numberOfMatches,
                                                                      highlightedResultIndex: activeMatchOrdinal,
                                                                      searchResultDisplayStyle: 2) // matches UIFindSession.SearchResultDisplayStyle.none
        }
        
        let arguments: [String : Any?] = [
            "activeMatchOrdinal": activeMatchOrdinal,
            "numberOfMatches": numberOfMatches,
            "isDoneCounting": isDoneCounting
        ]
        guard let channel = channel else { return }
        channel.invokeMethod("onFindResultReceived", arguments: arguments)
    }

    private func canDispatchCallbacks() -> Bool {
        guard channel != nil, let findInteractionController = findInteractionController else { return false }
        guard let webView = findInteractionController.webView else { return true }
        return webView.acceptsCallbacks()
    }
    
    public override func dispose() {
        super.dispose()
        findInteractionController = nil
    }
    
    deinit {
        dispose()
    }
}
